src/cinder/gl/GlslProg.cpp

```diff
@@ -457,573 +810,870 @@ std::string GlslProg::getShaderLog( GLuint handle ) const
-			mLoggedMissingUniforms.insert( name );
-		}
+	auto found = findUniform( lookUp );
+	if( ! found ) {
+		logMissingUniform( lookUp );
+		return;
 	}
-	else
-		glUniform4f( loc, data.r, data.g, data.b, data.a );
+	if( validateUniform( *found, data ) )
+		uniformMatFunc( found->mLoc, data, transpose );
 }
 
-// float*, count
-void GlslProg::uniform( int location, const float *data, int count ) const
+template<typename LookUp, typename T>
+inline void	GlslProg::uniformImpl( const LookUp &lookUp, const T *data, int count ) const
 {
-	ScopedGlslProg shaderBind( this );
-	glUniform1fv( location, count, data );
+	auto found = findUniform( lookUp );
+	if( ! found ) {
+		logMissingUniform( lookUp );
+		return;
+	}
+	if( validateUniform( *found, data, count ) )
+		uniformFunc( found->mLoc, data, count );
 }
 
-void GlslProg::uniform( const std::string &name, const float *data, int count ) const
+template<typename LookUp, typename T>
+inline void	GlslProg::uniformMatImpl( const LookUp &lookUp, const T *data, int count, bool transpose ) const
 {
-	ScopedGlslProg shaderBind( this );
-	GLint loc = getUniformLocation( name );
-	if( loc == -1 ) {
-		if( mLoggedMissingUniforms.count( name ) == 0 ) {
-			CI_LOG_E( "Unknown uniform: \"" << name << "\"" );
-			mLoggedMissingUniforms.insert( name );
-		}
+	auto found = findUniform( lookUp );
+	if( ! found ) {
+		logMissingUniform( lookUp );
+		return;
 	}
-	else
-		glUniform1fv( loc, count, data );
+	if( validateUniform( *found, data, count ) )
+		uniformMatFunc( found->mLoc, data, count, transpose );
 }
-
-// vec2*, count
-void GlslProg::uniform( int location, const vec2 *data, int count ) const
+	
+template<typename T>
+inline bool GlslProg::validateUniform( const Uniform &uniform, const T &val ) const
 {
-	ScopedGlslProg shaderBind( this );
-	glUniform2fv( location, count, &data[0].x );
+	if( ! checkUniformType<T>( uniform.mType ) ) {
+		logUniformWrongType( uniform.mName, uniform.mType, cppTypeToGlslTypeName<T>() );
+		return false;
+	}
+	else {
+		return checkUniformValue( uniform, &val, 1 );
+	}
 }
 
-void GlslProg::uniform( const std::string &name, const vec2 *data, int count ) const
+template<typename T>
+inline bool GlslProg::validateUniform( const Uniform &uniform, const T *val, int count ) const
 {
-	ScopedGlslProg shaderBind( this );
-	GLint loc = getUniformLocation( name );
-	if( loc == -1 ) {
-		if( mLoggedMissingUniforms.count( name ) == 0 ) {
-			CI_LOG_E( "Unknown uniform: \"" << name << "\"" );
-			mLoggedMissingUniforms.insert( name );
-		}
+	if( ! checkUniformType<T>( uniform.mType ) ) {
+		logUniformWrongType( uniform.mName, uniform.mType, cppTypeToGlslTypeName<T>() + "[" + to_string( count ) + "]" );
+		return false;
+	}
+	else {
+		return checkUniformValue( uniform, &val, 1 );
 	}
-	else
-		glUniform2fv( loc, count, &data[0].x );
 }
 
-// vec3*, count
-void GlslProg::uniform( int location, const vec3 *data, int count ) const
+template<typename T>
+string GlslProg::cppTypeToGlslTypeName()
```

Comment:
Actually, you're totally right! We started with a compile time check system. It was definitely messy. I think doing this would be fine. And it may be better passing type_info in for future proofing.